test/python/circuit/library/test_linear_function.py

```diff
@@ -0,0 +1,224 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Tests for LinearFunction class."""
+
+import unittest
+import numpy as np
+from ddt import ddt, data
+
+from qiskit.test import QiskitTestCase
+from qiskit.circuit import QuantumCircuit
+
+from qiskit.circuit.library.standard_gates import CXGate, SwapGate
+from qiskit.circuit.library.generalized_gates import LinearFunction
+from qiskit.circuit.exceptions import CircuitError
+
+from qiskit.quantum_info.operators import Operator
+
+
+def random_linear_circuit(num_qubits, num_gates, seed=None):
+    """Generate a pseudo random linear circuit."""
+
+    gates = ["cx", "swap"]
+
+    instructions = {
+        "cx": (CXGate(), 2),
+        "swap": (SwapGate(), 2),
+    }
+
+    if isinstance(seed, np.random.Generator):
+        rng = seed
+    else:
+        rng = np.random.default_rng(seed)
+
+    samples = rng.choice(gates, num_gates)
+
+    circ = QuantumCircuit(num_qubits)
+
+    for name in samples:
+        gate, nqargs = instructions[name]
+        qargs = rng.choice(range(num_qubits), nqargs, replace=False).tolist()
+        circ.append(gate, qargs)
+
+    return circ
+
+
+def random_invertible_binary_matrix(num_qubits, seed=None):
+    """Generates a random invertible n x n binary matrix."""
+
+    # This code is adapted from random_cnotdihedral
+    if isinstance(seed, np.random.Generator):
+        rng = seed
+    else:
+        rng = np.random.default_rng(seed)
+
+    det = 0
+    while np.allclose(det, 0) or np.allclose(det, 2):
+        binary_matrix = rng.integers(2, size=(num_qubits, num_qubits))
+        det = np.linalg.det(binary_matrix) % 2
```

Comment:
I was slightly concerned that this could take a really long time to run for large matrices, but I found (https://math.stackexchange.com/questions/54246/probability-that-a-random-binary-matrix-is-invertible) that the probability of success of this routine tends to about 30%, which is acceptable.

Reply:
Thanks, I did not know this. I have copied the code above from `operators/dihedral/random.py`.